csfieldguide/static/interactives/password-guesser/js/password-guesser.js

```diff
@@ -0,0 +1,138 @@
+var urlParameters = require('../../../js/third-party/url-parameters.js');
+
+const SHA256 = require('crypto-js/sha256');
+const Hex = require('crypto-js/enc-hex');
+const Passwords = require('./passwords.js');
+
+// List indices:
+// [name, the password or encryption , salt, [encryption with salt]]
+const NAME = 0;
+const HASH = 1;
+const SALT = 2;
+const HASHWITHSALT = 3;
+
+var isSalted;
+var allPasswords = [];
+
+$(document).ready(function() {
+  isSalted = !(urlParameters.getUrlParameter('salted') == "false"); // Default should be true
+  if (isSalted) {
+    $('#salt-header').removeClass('d-none');
+    $('#salt-input').removeClass('d-none');
+  }
+  $('#pg-password-salt').val('');
+  $('#pg-password-guess').val('');
+
+  setUpTable();
+
+  $('#pg-submit-button').click(checkGuess);
+
+  $('.pg-img').click(function() {
+    var salt = $(this).parent().html().trim().substring(0, 16); // Get the 16 character salt
+    $('#pg-password-salt').val(salt);
+  })
+});
+
+function setUpTable() {
+  hashLanguagePasswords();
+
+  var i = 0;
+  var j = 0;
+  var x = Passwords.LanguagePasswords.length;
+  while (i < x) {
+    addToTable(Passwords.LanguagePasswords[i], j);
+    i++;
+    j++;
+  }
+  i = 0;
+  x = Passwords.WordPasswords.length;
+  while (i < x) {
+    addToTable(Passwords.WordPasswords[i], j);
+    i++;
+    j++;
+  }
+}
+
+function hashLanguagePasswords() {
+  var password;
+  var hashWithSalt;
+  var hashNoSalt;
+  for (var i=0; i < Passwords.LanguagePasswords.length; i++) {
+    password = Passwords.LanguagePasswords[i];
+    hashWithSalt = SHA256(password[HASH] + password[SALT]).toString(Hex).toUpperCase();
+    hashNoSalt = SHA256(password[HASH]).toString(Hex).toUpperCase();
+    Passwords.LanguagePasswords[i] = [password[NAME], hashNoSalt, password[SALT], hashWithSalt];
+  }
+}
+
+function addToTable(password, index) {
+  var $table = $('#password-table');
+  console.log(index);
+  var html =  '<tr id="row-' + index + '">\n' +
+              '  <td>\n' +
+              '    ' + password[NAME] + '\n' +
+              '  </td>\n';
+  if (isSalted) {
+    html +=   '  <td class="salt">\n' +
+              '    ' + password[SALT] + '\n' +
+              '    <img class="pg-img" src="https://img.icons8.com/material-rounded/24/000000/copy.png">\n' +
+              '  </td>\n';
+  }
+  html +=     '  <td class="hash">\n' +
+              '    ' + (isSalted ? password[HASHWITHSALT] : password[HASH]) + '\n' +
+              '  </td>\n' +
+              '</tr>\n';
+  
+  allPasswords.push(password);
+  $table.append(html);
+}
+
+/**
+ * Concatenates the guessed password with the given salt, hashes the result,
+ * and checks it against hashed passwords on the table.
+ * 
+ * The salt is interpreted as a string (despite being hex) but there is no
+ * significant disadvantage to this
+ */
+function checkGuess() {
```

Comment:
It would be cool if the correctly guessed passwords remained highlighted so we knew which ones we had already guessed. Especially since one of the activities is to see which group can guess the most passwords.